lib: add navigator.deviceMemory

Uzlopak opened the pull request:
Shamelessly copied approximation logic from chromium

https://github.com/chromium/chromium/blob/main/third_party/blink/common/device_memory/approximated_device_memory.cc

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

H4ad commented on src/node_os.cc:
```diff
@@ -146,6 +146,40 @@ static void GetTotalMemory(const FunctionCallbackInfo<Value>& args) {
+  else
+    approximated_device_memory_gb_ = static_cast<float>(upper_bound) / 1024.0;
+
+  // Max-limit the reported value to 8GB to reduce fingerprintability of
+  // high-spec machines.
+  if (approximated_device_memory_gb_ > 8) approximated_device_memory_gb_ = 8.0;
```
I don't think finger-printability is an issue on Node, but I'm not sure about this, so raising this comment just to get some feedback.

At least on spec, is just a recommendation: https://www.w3.org/TR/device-memory/#computing-device-memory-value

targos replied:
Yeah, fingerprinting is not a concern in Node.js. Also, even if it was, why not use `os.totalmem()/1024**3` to get the value?

Uzlopak replied:
I just ensured that the code is in sync with chrome and according to the documentation.

benjamingr replied:
```suggestion
```

I prefer we remove this check

tniessen commented:
`navigator` is not a good API and I'm still unsure whether it was a good idea to add it at all. The only benefit is interoperability with other JavaScript runtimes, but it seems that neither Safari nor Firefox implement the `deviceMemory` property. I suppose Google's main goal is to establish the `Device-Memory` header, but that's unrelated to this PR and generally not very useful for server runtimes.

aduh95 commented:
+1 to what Tobias said, and also I'd be -1 on introducing properties to our implementation of `Navigator` unless it's listed on the HTML spec (https://html.spec.whatwg.org/multipage/system-state.html#the-navigator-object).

GeoffreyBooth commented:
Let’s please ping @nodejs/web-standards for all issues and PRs related to `navigator`.

ljharb commented on lib/internal/navigator.js:
```diff
@@ -24,7 +24,8 @@ const nodeVersion = process.version;
   #availableParallelism;
-  #deviceMemory;
+  #deviceMemory = getApproximatedDeviceMemory();
+  #userAgent = `Node.js/${nodeVersion.slice(1, nodeVersion.indexOf('.'))}`;
```
```suggestion
  #userAgent = `Node.js/${StringPrototypeSlice(nodeVersion, 1, StringPrototypeIndexOf(nodeVersion, '.'))}`;
```

Uzlopak replied:
created PR #50467 :)

jasnell commented:
Whether or not folks think `navigator` is a good API in general is not all that material here, in my opinion. I think the more relevant question is whether the new property yields any significant value. The main reason to add anything to `navigator` is cross-runtime interop and portability. The fact that none of the other runtimes expose `navigator.deviceMemory` and it is currently only implemented in blink-based browsers rather defeats that argument. I wouldn't say no to adding this but I also don't think there's value is adding it right now.

Uzlopak commented:
@jasnell 

To be honest, I think nodejs should consider itself as the "brother" of Chrome because both use v8. Bun uses the same engine as Safari, so their behavior is closer than it is to Chrome and nodejs. 

Thats why I also directly took the code from the chromium sourcecode, as I would prefer function parity with Chrome.

tniessen commented:
> To be honest, I think nodejs should consider itself as the "brother" of Chrome because both use v8. Bun uses the same engine as Safari, so their behavior is closer than it is to Chrome and nodejs. 
> 
> Thats why I also directly took the code from the chromium sourcecode, as I would prefer function parity with Chrome. 

Why should the choice of using V8 imply anything about feature parity between Node.js and Chrome? This API is not part of V8, not part of the JS language specification, and currently not even part of the HTML specification. Deno, Cloudflare Workers, etc. are also based on V8, and none of them have implemented this API as far as I know.

targos reviewed (changes requested):
Implementation is unnecessarily complex

benjamingr reviewed (approved):
LGTM with the fingerprinting check removed

Uzlopak commented:
@targos 
implemented it now in javascript according to the spec by using the most significant byte.

@benjamingr 
removed the lower and upper bounds.

@nodejs/web-standards 
PTAL

tniessen reviewed (changes requested):
I don't think this should land, for the reasons outlined in https://github.com/nodejs/node/pull/50229#issuecomment-1767878558.

I also don't understand @Uzlopak's argument that _"nodejs should consider itself as the "brother" of Chrome because both use v8"_ and that one might thus _"prefer function parity with Chrome."_

Uzlopak commented:
My thoughts were, as I stated before, that navigator.deviceMemory is a feature in Chrome. Just want to remind you that implementing the navigator global in node is about browser compatibility. Also this feature has a market share of 75% (chrome 63%, edge 5%, opera 3%, Samsung Internet 2%, etc..). 
Personally for me Chrome sets the standard and not Firefox with 3 % or Safari with 19% market share ...

And just because Firefox and Safari didnt implement it, doesnt result in the conclusion that we should not implement it - same as Chrome supporting this feature is not an argument for you.

A good argument against this feature is, that it is still a draft since 2018. Maybe there is a rule already covering these cases

navigator.deviceMemory is also used in few github projects https://github.com/search?q=navigator.deviceMemory&type=code 

So we could agree on not implementing a draft spec. Or we can agree on implemeting something which is supported by the most used browser.

I personally dont have a strong opinion on this feature. I just saw a low hanging fruit, so I picked it.

panva reviewed (changes requested):
> Just want to remind you that implementing the navigator global in node is about browser compatibility

I would argue the point is interoperability, not 100% compatibility.

Since this is a) draft, b) Chrome-only property and so c) one already has to work around its lack of presence in other runtimes if they wish to have interoperable code, I don't think we should be landing this.

Uzlopak commented:
Allrighty. :)
